File: structures/binary_tree/binary_tree.cpp
#include <iostream>
#include <string>
#include <stack>
#include <stdlib.h>
#include <queue>

using std::string;
using std::cout;
using std::cin;
using std::endl;
using std::stack;
using std::queue;

//структура, описывающая узел в дереве выражений
struct Node
{
    char value;
    Node* left;
    Node* right;

    Node(char str) {value = str; left = right = nullptr; }
};  

//структура, описывающая узел в дереве поиска
struct Node_Sr
{
    int value;
    Node_Sr* left;
    Node_Sr* right;

    Node_Sr(int val) {value = val; left = right = nullptr; }
    Node_Sr() {value = 0; left = right = nullptr; }
};  

//дерево выражений
class ExpTree
{
public:
    Node* root; //корень

public:
    //создаёт представление дерева в виде стека
    Node* build_tree(const string& str)
    {
        stack<Node*> stack; //т.к. при постфиксной записи вершина добавляется в конце, то стек идеально подходит 
        
        for (char l : str)
        {
            if (l == ' ') continue; //если пробел

            //если это буква или число  
            if (isalnum(l)) stack.push(new Node(l));
            else // иначе это опреция: + - и т.д.
            {
                //достаем два последних значения, т.к. бинарное дерево
                Node* right = stack.top(); stack.pop();
                Node* left = stack.top(); stack.pop();
                //образуем новый узел
                Node* node = new Node(l);
                node->left = left;
                node->right = right;
                //добавляем узел в стек 
                stack.push(node);
            }
        }

        return stack.top(); //по сути указатель на вершину дерева
    }

    ExpTree(const string& str) {root = build_tree(str);} //конструктор
    ~ExpTree() {clear(root);} //диструктор

    //очистка памяти
    void clear(Node* node)
    {
        if (node == nullptr) return;
        clear(node -> left);
        clear(node -> right);
        delete node;
    }

    //постфиксный обход (лево, право, узел)
    void post(Node* node)
    {
        if (node == nullptr) return;
        post(node -> left);
        post(node -> right);
        cout << node -> value << " ";
    }

    //префиксный обход (узел, лево, право)
    void pred(Node* node)
    {
        if (node == nullptr) return;
        cout << node -> value << " ";
        pred(node -> left);
        pred(node -> right);
    }

    //симметричный обход (лево узел право)
    void symm(Node* node)
    {
        if (node == nullptr) return;
        symm(node -> left);
        cout << node -> value << " ";
        symm(node -> right);
    }

    //вывод дерева по уровням 
    void show_tree()
    {
        cout << "\nThe expression tree" << endl;
        if (root == nullptr) return;

        queue<Node*> q; //очередь для хранения узлов на текущем уровне
        q.push(root);
        
        int level = 0;
        
        while (!q.empty())
        {
            int size = q.size();  //сколько узлов на текущем уровне
            
            cout << "Level " << level << ": ";
            
            //обрабатываем все узлы текущего уровня
            for (int i = 0; i < size; i++)
            {
                Node* cur = q.front(); q.pop();
                cout << cur -> value << " ";
                
                if (cur -> left) q.push(cur -> left);
                if (cur -> right) q.push(cur -> right);
            }
            
            cout << endl;
            level++;
        }
        putchar('\n');
    }
};

//дерево поиска
class SearchTree
{
public:
    Node_Sr* root;

public:
    SearchTree() {root = nullptr;}
    ~SearchTree() {clear(root);}

    //очистка памяти
    void clear(Node_Sr* node)
    {
        if (node == nullptr) return;
        clear(node -> left);
        clear(node -> right);
        delete node;
    }

    //добавление элемента
    Node_Sr* append(Node_Sr* node, int val)
    {
        //если первый элемент
        if (node == nullptr) return new Node_Sr(val);

        //если дубль
        else if (node -> value == val) return nullptr;

        //если значение в текущей вершине больше, то идём влево
        else if (node -> value > val) node -> left = append(node -> left, val);

        //если значение в текущей вершине меньше, то идём вправо
        else if (node -> value < val) node -> right = append(node -> right, val);

        return node;
    }

    //поиск листа слева, т.е. мин значения
    Node_Sr* find_min(Node_Sr* node)
    {
        while (node && node -> left) node = node -> left;
        return node;
    }

    //удаление узла, ветки или листа
    Node_Sr* del_Node(Node_Sr* node, int key)
    {   
        if (node == nullptr) return node;

        //если ключ меньше ищем слева, иначе справа 
        if (key < node -> value) node -> left = del_Node(node -> left, key);
        else if (key > node -> value) node->right = del_Node(node -> right, key);
        
        //если нашли 
        else 
        {
            //если лист
            if (node -> left == nullptr && node -> right == nullptr)
            {
                delete node; return nullptr;
            }

            //если только 1 потомок
            else if (node -> left == nullptr || node -> right == nullptr)
            {
                Node_Sr* tmp;
                if (node -> left) tmp = node -> left;
                if (node -> right) tmp = node -> right;
                delete node;
                return tmp;
            }

            //2 потомка
            else
            {
                Node_Sr* t = find_min(node -> right); //находим минимальное значение в правой ветке
                node -> value = t -> value; //присваиваем удаляемой вершине мин значение
                node -> right = del_Node(node -> right, t -> value); //удаляем вершину с взятым мин значением
            }
        }

        return node;
    }

    //вывод дерева - постфиксный обход
    void show_tree(Node_Sr* node)
    {
        if (node == nullptr) return;
        show_tree(node -> left);
        show_tree(node -> right);
        cout << node -> value << " ";
    }

    //вывод дерева в ширину
    void show_width_tree(Node_Sr* node)
    {
        cout << "\nThe search tree" << endl;
        if (node == nullptr) return;

        queue<Node_Sr*> q;
        int level = 0;
        q.push(node);

        while (q.empty() == 0)
        {
            int size = q.size();
            cout << "Level " << level << ": ";

            for (int i = 0; i < size; i++)
            {
                Node_Sr* cur = q.front(); q.pop();
                cout << cur -> value << " ";

                if (cur -> left) q.push(cur -> left);
                if (cur -> right) q.push(cur -> right);
            }

            putchar('\n');
            level++;
        }
    }
};

int main()
{
    ExpTree* tree = new ExpTree("ADE-BC-+F*G-H/-");
    tree -> show_tree();
    cout << "--------" << endl;
    cout << "\npost" << endl;
    tree -> post(tree -> root);
    cout << "\n--------" << endl;
    cout << "\npred" << endl;
    tree -> pred(tree -> root);
    cout << "\n--------" << endl;
    cout << "\nsymm" << endl;
    tree -> symm(tree -> root);
    cout << "\n--------" << endl;

    SearchTree* sr_tree = new SearchTree();
    const int size = 13;
    int nums[size] = {5, 16, 26, 3, 20, 29, 54, 37, 34, 48, 40, 10, 43};

    for (int i = 0; i < size; i++ )sr_tree -> root = sr_tree -> append(sr_tree -> root, nums[i]);
    sr_tree -> show_tree(sr_tree -> root);
    cout << "\n------" << endl;
    sr_tree -> show_width_tree(sr_tree -> root);
    cout << "\n------" << endl;
    sr_tree -> del_Node(sr_tree -> root, 5);
    cout << "Delete elem with number 5" << endl;
    sr_tree -> show_width_tree(sr_tree -> root);
    cout << "\n------" << endl;
    sr_tree -> show_tree(sr_tree -> root);
    cout << "\n------" << endl;


    return 0;
}

File: structures/binary_tree/binary_tree.py
class Node:
    def __init__(self, data):
        self.data = data #значение
        self.left = self.right = None #ветки

class Tree:
    def __init__(self):
        self.root = None #вершина

    #метод для поиска указанного значения в вершинах 
    def find(self, node, parent, value):
        if node is None: return None, parent, False #если вершины не существует

        if value == node.data: return node, parent, True #если нашли значение в дереве

        if value < node.data:
            if node.left: return self.find(node.left, node, value) #если на текущем уровне нет значения, и это значение меньше чем в текущей вершине, то ищем слева

        if value > node.data:
            if node.right: return self.find(node.right, node, value) #если на текущем уровне нет значения, и это значение больше чем в текущей вершине, то ищем справа

        return node, parent, False #если значение не найдено в дереве
    
    #добавление вершины
    def append(self, obj):
        #если это не объект класса Node
        if not isinstance(obj, Node): return None

        # если это первая вершина в дереве 
        if self.root is None:
            self.root = obj
            return obj
        
        #находим свободную вершину, её родителя и проверяем, чтобы нового значения не было в дереве
        cur, parent, flag = self.find(self.root, None, obj.data)
        
        #если дубликата нет и вершина, после которой мы хотим записать новое значение существует
        if flag == False and cur:
            if obj.data < cur.data: cur.left = obj
            else: cur.right = obj;

        return obj
    
    def show(self, node):
        if node is None: return None

        self.show(node.left)
        print(node.data, end=' ')
        self.show(node.right)

    def show_wide_tree(self, node):
        #если вершины, от которой мы хотим отобразить дерево, нет
        if node is None: return node

        v = [node] #список с текущими вершинами
        level = flag = 0
        while v:
            buf = []
            for top in v:
                if (flag == 0):
                    print(f"Level tree: {level}.", top.data, end=" ")
                    flag = 1
                else: print(top.data, end=" ")

                if top.left: buf += [top.left]
                if top.right: buf += [top.right]

            level += 1
            flag = 0
            print()
            v = buf

    #удаление листка 
    def del_leaf(self, cur, parent):
        if parent.left == cur: parent.left = None
        else: parent.right = None

    #удаление вершины
    def del_top(self, cur, parent):
        if parent.left == cur:
            if cur.right: parent.left = cur.right
            elif cur.left: parent.left = cur.left

        elif parent.right == cur:
            if cur.right: parent.right = cur.right
            elif cur.left: parent.right = cur.left

    #ищет минимальное значение в текущей ветке
    def find_min(self, node, parent):
        #наим. значение всегда будет в левом узле, поэтому пока он существует у текущей вершины, то мы идёт по нему
        if node.left: return self.find_min(node.left, node)

        return node, parent
    
    #удаление ветки
    def del_node(self, key):
        cur, parent, flag = self.find(self.root, None, key)

        #если значение key не найдено
        if not flag: return None 

        if isinstance(cur, Node):
            #если оба дочерних узло не опрделены, то это лист
            if cur.right is None and cur.left is None: self.del_leaf(cur, parent)

            #если только одна вершина имеет потомка, то удаляем как вершину
            elif cur.right is None or cur.left is None: self.del_top(cur, parent)   

            else:
                #если нужно удалить целый узел, то находим минимальное значение в правой ветке, записываем его в удаляемую вершину и удаляем вершину с этим мин. значением                cur_right, parent_right = self.find_min(cur, parent)
                cur_right, parent_right = self.find_min(cur, parent)
                cur.data = cur_right.data
                self.del_top(cur_right, parent_right)

v = [10, 5, 7, 16, 13, 2, 20]
#v = [20, 5, 24, 2, 16, 11, 18]
 
t = Tree()
for x in v:
    t.append(Node(x))
 
# t.del_node(5)
t.show_wide_tree(t.root)
t.show(t.root)